src/lib/pubkey/pcurves/pcurves_util.h

```diff
@@ -0,0 +1,188 @@
+/*
+* (C) 2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_PCURVES_UTIL_H_
+#define BOTAN_PCURVES_UTIL_H_
+
+#include <botan/internal/mp_core.h>
+#include <array>
+
+namespace Botan {
+
+template <WordType W, size_t N, size_t XN>
+inline consteval std::array<W, N> reduce_mod(const std::array<W, XN>& x, const std::array<W, N>& p) {
+   std::array<W, N + 1> r = {0};
+   std::array<W, N + 1> t = {0};
+
+   const size_t x_bits = XN * WordInfo<W>::bits;
+
+   for(size_t i = 0; i != x_bits; ++i) {
+      const size_t b = x_bits - 1 - i;
+
+      const size_t b_word = b / WordInfo<W>::bits;
+      const size_t b_bit = b % WordInfo<W>::bits;
+      const bool x_b = (x[b_word] >> b_bit) & 1;
+
+      shift_left<1>(r);
+      if(x_b) {
+         r[0] += 1;
+      }
+
+      const W carry = bigint_sub3(t.data(), r.data(), N + 1, p.data(), N);
+
+      if(carry == 0) {
+         std::swap(r, t);
+      }
+   }
+
+   std::array<W, N> rs;
+   std::copy(r.begin(), r.begin() + N, rs.begin());
```

Comment:
Now that `copy_mem` has a constexpr-variant, I think we should prefer that. If both passed-in ranges have static length, it will do a `static_assert` of the correct lengths.

```suggestion
   std::copy(std::span{r}.first<N>(), rs);
```

Reply:
:+1: